packages/block-library/src/image/edit.js

```diff
@@ -170,10 +195,30 @@ class ImageEdit extends Component {
 			isEditing: false,
 		} );
 
+		const blockWidth = getBlockWidth();
+		let src = media.url;
+		let img = {};
+		let actualWidth;
+		let actualHeight;
+
+		if ( media.sizes ) {
+			// The "full" size is already included in `sizes`.
+			img = media.sizes.large || media.sizes.full;
+			src = img.url;
+			actualWidth = get( img, [ 'actual_size', 'width' ] ) || img.width;
+			actualHeight = get( img, [ 'actual_size', 'height' ] ) || img.height;
+		}
+
+		const attr = pickRelevantMediaFiles( media );
+		attr.url = src;
+
 		this.props.setAttributes( {
-			...pickRelevantMediaFiles( media ),
-			width: undefined,
-			height: undefined,
+			...attr,
+
+			// Not used in the editor, passed to the front-end in block attributes.
+			fileWidth: actualWidth,
+			fileHeight: actualHeight,
+			editWidth: blockWidth,
```

Comment:
I'm finding some of the new mutation difficult to follow - to see what it's supposed to be setting. I _think_ we're wanting to set the actual image size if given, else use the default sizes as if none is given?

```js
const { sizes: givenSizes } = media;
const givenSize = givenSizes && ( givenSizes.large || givenSizes.full );

this.props.setAttributes( Object.assign(
	{
		url: media.url,
		editWidth: getBlockWidth(),
	},
	pickRelevantMediaFiles( media ),
	givenSizes && {
		url: size.url,
		fileWidth: get( givenSize, 'actual_size.width', givenSize.width ),
		fileHeight: get( givenSize, 'actual_size.height', givenSize.height ),
	}
} );
```

Reply:
The whole problem is that the image meta in the media modal is "tinted". The sizes are passed through [image_constrain_size_for_editor()](https://core.trac.wordpress.org/browser/tags/4.9.8/src/wp-includes/media.php#L27) which uses the global `$content_width` to limit them. However we want the "normal" sizes, so had to add https://github.com/WordPress/gutenberg/pull/11973/files/03444d7bdda0d79dcc540d0fbfa5141c2a14de78#diff-6ff32417da0658502e7caa1a1abbeae6R309 to fix it.

It's not really a mutation, it tries to pick one value if it exists, and falls back to another no matter what it is. Can be written in several ways, the wordier one seems more readable :)